ding/model/template/ebm.py

```diff
@@ -0,0 +1,497 @@
+        probs = F.softmax(-1.0 * energies, dim=-1)
+        # (B, )
+        best_idxs = probs.argmax(dim=-1)
+        return action_samples[torch.arange(action_samples.size(0)), best_idxs]
+
+    def set_action_bounds(self, action_bounds: np.ndarray):
+        self.action_bounds = torch.as_tensor(action_bounds, dtype=torch.float32).to(self.device)
+
+    @abstractmethod
+    def sample(self, obs: torch.Tensor, ebm: nn.Module) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Create tiled observations and sample counter-negatives for feeding to the InfoNCE objective.
+        obs: (B, O)
+        return: (B, N, O), (B, N, A).
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def infer(self, obs: torch.Tensor, ebm: nn.Module) -> torch.Tensor:
+        """Optimize for the best action conditioned on the current observation.
+        obs: (B, O)
+        return: (B, A).
+        """
+        raise NotImplementedError
+
+
+@STOCHASTIC_OPTIMIZER_REGISTRY.register('dfo')
+class DFO(StochasticOptimizer):
+    def __init__(
+        self,
+        noise_scale: float = 0.33,
+        noise_shrink: float = 0.5,
+        iters: int = 3,
+        train_samples: int = 8,
+        inference_samples: int = 16384,
+        cuda: bool = False,
+    ):
+        self.action_bounds = None
+        self.noise_scale = noise_scale
+        self.noise_shrink = noise_shrink
+        self.iters = iters
+        self.train_samples = train_samples
+        self.inference_samples = inference_samples
+        self.device = torch.device('cuda' if cuda else "cpu")
+
+    def sample(self, obs: torch.Tensor, ebm: nn.Module) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Create tiled observations and sample counter-negatives for feeding to the InfoNCE objective.
+        obs: (B, O)
+        return: (B, N, O), (B, N, A).
+        """
+        del ebm
+        return self._sample(obs, self.train_samples)
+
+    @torch.no_grad()
+    def infer(self, obs: torch.Tensor, ebm: nn.Module) -> torch.Tensor:
+        """Optimize for the best action conditioned on the current observation.
+        obs: (B, O)
+        return: (B, A).
+        """
+        noise_scale = self.noise_scale
+
+        # (B, N, O), (B, N, A)
+        obs, action_samples = self._sample(obs, self.inference_samples)
+
+        for i in range(self.iters):
+            # (B, N)
+            energies = ebm.forward(obs, action_samples)
+            probs = F.softmax(-1.0 * energies, dim=-1)
+
+            # Resample with replacement.
+            idxs = torch.multinomial(probs, self.inference_samples, replacement=True)
+            action_samples = action_samples[torch.arange(action_samples.size(0)).unsqueeze(-1), idxs]
+
+            # Add noise and clip to target bounds.
+            action_samples = action_samples + torch.randn_like(action_samples) * noise_scale
+            action_samples = action_samples.clamp(min=self.action_bounds[0, :], max=self.action_bounds[1, :])
```

Comment:
self.action_bounds[0, :] maybe a tensor containing more than one numbers

Reply:
lets say  `action_sample=[[1, 2, 3],[1, 2, 3]]` where the first dimension is the batch dimension.
and `self.action_bounds[0, :]=[0, 2, 4]`,
then `action_samples.clamp(max=self.action_bounds[0, :]) = [[0, 2, 3], [0, 2, 3]]`

Reply:
Got it.